runtime/parachains/src/paras.rs

```diff
@@ -0,0 +1,1130 @@
+
+			// the candidate is in the context of the parent of `applied_after`,
+			// thus does not trigger the code upgrade.
+			{
+				Paras::note_new_head(para_id, Default::default(), applied_after - 1);
+
+				assert!(Paras::past_code_meta(&para_id).most_recent_change().is_none());
+				assert_eq!(<Paras as Store>::FutureCodeUpgrades::get(&para_id), Some(applied_after));
+				assert_eq!(<Paras as Store>::FutureCode::get(&para_id), new_code);
+				assert_eq!(Paras::current_code(&para_id), Some(vec![1, 2, 3].into()));
+			}
+
+			run_to_block(applied_after + 1, None);
+
+			// the candidate is in the context of `applied_after`, and triggers
+			// the upgrade.
+			{
+				Paras::note_new_head(para_id, Default::default(), applied_after);
+
+				assert_eq!(
+					Paras::past_code_meta(&para_id).most_recent_change(),
+					Some(applied_after),
+				);
+				assert_eq!(
+					<Paras as Store>::PastCode::get(&(para_id, applied_after)),
+					Some(vec![1, 2, 3,].into()),
+				);
+				assert!(<Paras as Store>::FutureCodeUpgrades::get(&para_id).is_none());
+				assert!(<Paras as Store>::FutureCode::get(&para_id).0.is_empty());
+				assert_eq!(Paras::current_code(&para_id), Some(new_code));
+			}
+		});
+	}
+
+	#[test]
+	fn code_upgrade_applied_after_delay_even_when_late() {
+		let acceptance_period = 10;
+		let validation_upgrade_delay = 5;
+
+		let paras = vec![
+			(0u32.into(), ParaGenesisArgs {
+				parachain: true,
+				genesis_head: Default::default(),
+				validation_code: vec![1, 2, 3].into(),
+			}),
+		];
+
+		let genesis_config = MockGenesisConfig {
+			paras: GenesisConfig { paras, ..Default::default() },
+			configuration: crate::configuration::GenesisConfig {
+				config: HostConfiguration {
+					acceptance_period,
+					validation_upgrade_delay,
+					..Default::default()
+				},
+				..Default::default()
+			},
+			..Default::default()
+		};
+
+		new_test_ext(genesis_config).execute_with(|| {
+			let para_id = ParaId::from(0);
+			let new_code = ValidationCode(vec![4, 5, 6]);
+
+			run_to_block(2, None);
+			assert_eq!(Paras::current_code(&para_id), Some(vec![1, 2, 3].into()));
+
+			let applied_after = {
+				// this parablock is in the context of block 1.
+				let applied_after = 1 + validation_upgrade_delay;
+				Paras::schedule_code_upgrade(para_id, new_code.clone(), applied_after);
+				Paras::note_new_head(para_id, Default::default(), 1);
+
+				assert!(Paras::past_code_meta(&para_id).most_recent_change().is_none());
+				assert_eq!(<Paras as Store>::FutureCodeUpgrades::get(&para_id), Some(applied_after));
+				assert_eq!(<Paras as Store>::FutureCode::get(&para_id), new_code);
+				assert_eq!(Paras::current_code(&para_id), Some(vec![1, 2, 3].into()));
+
+				applied_after
+			};
+
+			run_to_block(applied_after + 1 + 4, None);
+
+			// the candidate is in the context of the first descendent of `applied_after`, and triggers
+			// the upgrade.
+			{
+				Paras::note_new_head(para_id, Default::default(), applied_after + 4);
```

Comment:
Again, not that easy to understand these consts.

A side note: since we are already adding ascii schemes here and there probably some with the timelines of what's going on in the testcases might be helpful:

```
  <  [1, 2, 3] lives here  > < upgrade to [4, 5, 5]   >
 +--------------------------+--------------------------+--------------------------> time

  <  validation upgrade    >
          delay

```

Reply:
The reason is the same as the above, just with an added delay of `4`.